### DISCUSSION TITLE
Correct way to handle async IO inside `register_callback` within Tokio runtime

### DISCUSSION THREAD

**Discussion by 0xForerunner** [open-telemetry/opentelemetry-rust#1362] [Q&A]
I would like to send http requests during the callback. Should I use spawn_blocking for the entire pipeline? Or is it okay to having blocking io within the callback?

**Comment by jtescher** [comment]
you can block for a reasonable amount of time within the callback, we can consider making the callback accept an async closure too.

**Reply by 0xForerunner** [reply]
Okay cool good to know! I think something like `register_callback_async` would be pretty handy I'm sure!

**Reply by 0xForerunner** [reply]
So after some investigation I have found that to not be true. This code sadly causes a deadlock.
```rust
    meter.register_callback(&anys, move |observer| {
        println!("In callback");
        block_on(metrics_callback(observer)).log();
    })?;
```
Not really sure what the solution is.

**Reply by 0xForerunner** [reply]
Although now that I think of it this might be happening because I am accessing a static RwLock inside that function and it doesn't like block_on.... hmmm

**Reply by 0xForerunner** [reply]
So it's not from the lazy static or the RwLock. it seems something from
```rust
client.query().await
```
is causing the deadlock. Not sure why this would be.

**Comment by GabrielBianconi** [comment]
For future reference: 

Here's a hacky solution I found for this issue:

https://github.com/open-telemetry/opentelemetry-rust/issues/1376#issuecomment-1987102217
